lib/gps.js

```diff
@@ -0,0 +1,394 @@
+
+
+/**
+ *
+ * @constructor
+ *
+ * @param {Object} opts Options: pin(s), chip, receiver, breakout, precision, serialport, freq
+ *
+ * Sample initialization
+ *
+ *    new five.GPS({ pins: {rx: 10, tx: 11});
+ *
+ */
+
+function GPS(opts) {
+
+  var breakout, receiver, chip, state;
+
+  if (!(this instanceof GPS)) {
+    return new GPS(opts);
+  }
+
+  // Allow users to pass in a 2 element array for tx and rx pins
+  if (Array.isArray(opts)) {
+    opts = { pins: {tx: opts[0], rx: opts[1], onOff: opts[2] }};
+  }
+
+  if (typeof opts.pins === "undefined") {
+    opts.pins = {};
+  }
+
+  Board.Component.call(
+    this, this.opts = Board.Options(opts)
+  );
+
+  breakout = opts.breakout || "adafruitUltimateGPS";
+
+  // Get user overrides for chip and receivers
+  receiver = opts.receiver;
+  chip = opts.chip;
+
+  if (!receiver && Breakouts[breakout].receiver) {
+    receiver = Breakouts[breakout].receiver.value;
+  }
+
+  if (!chip && Breakouts[breakout].chip) {
+    chip = Breakouts[breakout].chip.value;
+  }
+
+  // If a receiver was defined or derived but chip was not
+  if (receiver && !chip && Receivers[receiver].chip) {
+    chip = Receivers[receiver].chip.value;
+  }
+
+  // Allow users to pass in custom chip types
+  chip = typeof chip === "string" ?
+    Chips[chip] : this.opts.chip;
+
+  // Allow users to pass in custom receiver types
+  receiver = typeof receiver === "string" ?
+    Receivers[receiver] : this.opts.receiver;
+
+  // Chip decorates the instance
+  Object.defineProperties(this, chip);
+
+  // Receiver decorates this instance
+  Object.defineProperties(this, receiver);
+
+  // breakout decorates the instance
+  if (this.opts.breakout) {
+    breakout = typeof this.opts.breakout === "string" ?
+      Breakouts[this.opts.breakout] : this.opts.breakout;
+
+    Board.Controller.call(this, breakout, opts);
+  }
+
+  // If necessary set default property values
+  this.precision = opts.precision || 9;
+  this.baud = opts.baud || breakout.baud || chip.baud;
+
+  // Create a "state" entry for privately
+  // storing the state of the instance
+  state = {
+    sat: {},
+    pos: {},
+    nav: {},
+    frequency: 1000,
+    lowPowerMode: false
+  };
+
+  priv.set(this, state);
+
+  // Getters for private state values
+  Object.defineProperties(this, {
+    pos: {
+      get: function() {
+        return state.pos;
+      }
+    },
+    sat: {
+      get: function() {
+        return state.sat;
+      }
+    },
+    nav: {
+      get: function() {
+        return state.nav;
+      }
+    },
+    time: {
+      get: function() {
+        return state.time;
+      }
+    }
+  });
+
+  if (this.initialize) {
+    this.initialize(opts);
+  }
+
+}
+
+util.inherits(GPS, events.EventEmitter);
+
+/*
+ * Default intialization for serial GPS
+ */
+GPS.prototype.initialize = function(opts) {
+
+  var state = priv.get(this);
+  state.serialPort = opts.serialPort || opts.portId || opts.port || opts.bus || this.io.SERIAL_PORT_IDs.SW_SERIAL0;
```

Comment:
When defaulting to SW serial, it may also be wise to recommend specific pins. There are all sorts of tricky issues with SW serial because it relies on pin change interrupts. For example is you specify SW serial pins on the same port (the AVR port - not the Firmata port) that you are using for any digital input or output, it will mess up the SW serial transmission but not throw an error. Super frustrating to debug if the user doesn't understand the cause.